### ISSUE TITLE
dired-collapse-mode and dired-hide-details-mode

### ISSUE THREAD

**Issue by vkazanov** [Fuco1/dired-hacks#87]
dired-collapse-mode is just brilliant, thank you very much!

Unfortunately, it does not really interact with dired-hide-details-mode really well: the replaced string contains all the file attributes dired-hide-details-mode normally hides. 

It'd be great it fix the issue as dired-hide-details-mode, AFAIK, is quite popular.

**Comment by Fuco1** [comment]
Hm, and what happens when you toggle `dired-hide` off and on again?  If that fixes the problem then I guess we can just add this toggle to the hook and that's it.

**Comment by vkazanov** [comment]
To fix things visually I have to: switch off the `dired-hide-details-mode`, press `g` to reload the buffer, switch on `dired-hide-details-mode` again, enjoy both the hidden file details and collapsed directories.

I'll try to find a better solution or at least debug things a bit.

**Comment by vkazanov** [comment]
So, the following happens: hide-details-mode does not rewrite or insert anything, it just adds a property `dired-hide-details-detail` to invisibility-spec, i.e. the text is there but hidden.

`dired-collapse` needs some column position data to insert the rewritten text into the buffer, which it gets in `dired-collapse--get-column-info`. Because most of the columns are hidden the column data returned is broken and the collapsing function errors right before setting text properties.

**Comment by vkazanov** [comment]
The fast and stupid hack would be to wrap the whole `dired-collapse` body in disabling/enabling the hide-details-mode calls.
 
A cleaner way would be to somehow ignore text invisibility in dired buffers, or *locally* ignoring them.

**Comment by Fuco1** [comment]
Oh cool, thanks for the analysis.

You can indeed tweak the invisibility buffer-locally and temporarily at the same time, check the variable `buffer-invisibility-spec`.  If we let-bind this with proper value around the body of that function it would turn of the `dired-hide-details` invisibility and everything should work.

Do you want to try to make this work?  If so feel free to open a PR and we can discuss the code/solutions or I can help you if you get stuck.

**Comment by vkazanov** [comment]
Here's the PR. By the way, how do you enable the mode globally? So that Emacs would always open directories with the mod enabled and working already. Trivially enabling the mode in the `dired-mode-hook` does not quite work as the user will still have to `revert-buffer` for the mode to start working initially.

**Comment by Fuco1** [comment]
@vkazanov Heh, I haven't thought about that, I actually haven't had time yet to try to configure it :D

A lazy way would be to add it to the hook globally

```elisp
(add-hook 'dired-after-readin-hook 'dired-collapse 'append)
```

I should add some instructions.  However, when you enable the mode it should also automatically revert the buffer, does it not?

**Comment by vkazanov** [comment]
Nope, it doesn't. AFAIK, initially the hook gets called after dired has already read the directory listing.

**Comment by vkazanov** [comment]
Maybe it's reasonable to have something about that in the README itself?

**Comment by Fuco1** [comment]
> Nope, it doesn't. AFAIK, initially the hook gets called after dired has already read the directory listing.

That's true, but then in the definition of the mode I call `revert-buffer`.  So in the worst case, it should "read-in" twice.  Which I admit is not ideal, I haven't thought about it.

```elisp
(define-minor-mode dired-collapse-mode
  "Toggle collapsing of unique nested paths in Dired."
  :group 'dired-collapse
  :lighter ""
  (if dired-collapse-mode
      (add-hook 'dired-after-readin-hook 'dired-collapse 'append 'local)
    (remove-hook 'dired-after-readin-hook 'dired-collapse 'local))
  (revert-buffer))
```

**Comment by vkazanov** [comment]
Yes, I saw the code.

Unfortunately, this doesn't work the first time the buffer gets opened. I didn't yet look into it seriously, to be honest. Obviously, it has something to do with the order of hooks.

**Comment by Fuco1** [comment]
Oh, right.  It might happen that the local hooks run before global hooks (or the other way around) and that can screw with some other functionality.

This is one thing that's kind of sad about hooks in that you can't really control the dependencies very well.  Of course, ideally one would write code independent of order but when there are multiple extensions operating on the same "data" it just can't be avoided.

In the past I had a hack macro I would wrap around the hook body which would do something like: check the hook variable, and if the *current* hook depends on some later hook, do not run myself, but remove myself and append to the end.  You can imagine how well this worked out :D  But then I didn't know much elisp back then... maybe something like this could be created in a way that wouldn't be totally insane.
